### PR TITLE
feat(providers): Add support for fetching from k8s secrets

### PR DESCRIPTION
**Pull request by bersalazar**
Allows fetching values from Kubernetes secrets.

Authentication to the Kubernetes cluster is managed by referencing the local kubeconfig.

Covers #185

### CODE REVIEW

File: pkg/providers/k8s/k8s.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func New(l *log.Logger, cfg api.StaticConfig) *provider {
+	p := &provider{
+		log: l,
+	}
+
+	kubeConfig, err := getKubeConfig(cfg, *l)
+	if err != nil {
+		fmt.Printf("An error occurred getting the Kubeconfig path: %s", err)
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go]: only print?

Comment (bersalazar) [reply pkg/providers/k8s/k8s.go]: Thanks! It's a good point. It made me realize some improvements in the error handling that I've now pushed.

File: pkg/providers/k8s/k8s.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	secretName := splits[1]
+	key := splits[2]
+
+	if p.KubeConfigPath == "" {
+		return "", fmt.Errorf("No Kubeconfig path was found")
+	}
+
+	secretData, err := getSecret(namespace, secretName, p.KubeConfigPath, p.KubeContext, context.Background())
+	secret, exists := secretData[key]
+	if err != nil || !exists {
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go]: hmmm... we should handle it separately. it will be better.

**Comment by yxxhero** [comment]
@bersalazar could you add some tests for this feature?

**Comment by aslafy-z** [comment]
What about config maps? Maybe we can prep to support them in the future by leaving room in the api for them?

File: pkg/providers/k8s/k8s.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+func New(l *log.Logger, cfg api.StaticConfig) *provider {
+	p := &provider{
+		log: l,
+	}
+
+	kubeConfig, err := getKubeConfig(cfg)
+	if err != nil {
+		fmt.Printf("An error occurred getting the Kubeconfig path: %s\n", err)
+		return p
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go]: what will happen after return p without the right args?

File: pkg/providers/k8s/k8s.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+func getKubeConfig(cfg api.StaticConfig) (string, error) {
+	// Use kubeConfigPath from URI parameters if specified
+	if cfg.String("kubeConfigPath") != "" {
+		if _, err := os.Stat(cfg.String("kubeConfigPath")); err != nil {
+			return cfg.String("kubeConfigPath"), fmt.Errorf("kubeConfigPath URI parameter is set but path %s does not exist.", cfg.String("kubeConfigPath"))
+		}
+	}
+
+	// Use path in KUBECONFIG environment variable if set
+	if envPath := os.Getenv("KUBECONFIG"); envPath != "" {
```
Comment (aslafy-z) [inline pkg/providers/k8s/k8s.go:51]: The kubectl KUBECONFIG env variable impl supports a list of paths
https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#the-kubeconfig-environment-variable
I think we should support the same pattern here

File: pkg/providers/k8s/k8s.go (hunk 3 above)
Comment (bersalazar) [reply pkg/providers/k8s/k8s.go]: Got it. I was handling it at GetString() but now I return an error here if the kubeconfig fetch failed.

**Comment by bersalazar** [comment]
> What about config maps? Maybe we can prep to support them in the future by leaving room in the api for them?
> 
> Maybe /[namespace/][apiVersion/kind/]name[#path.to.key]

Definitely worth considering. Maybe in a future PR, happy to raise an issue for it once this one's merged.

File: pkg/providers/k8s/k8s.go (hunk 4 above)
Comment (bersalazar) [reply pkg/providers/k8s/k8s.go:51]: I can create an issue for it and see to picking it up at a later time if that's alright.

**Comment by aslafy-z** [comment]
> > What about config maps? Maybe we can prep to support them in the future by leaving room in the api for them?
> > 
> > Maybe /[namespace/][apiVersion/kind/]name[#path.to.key]
> 
> Definitely worth considering. Maybe in a future PR, happy to raise an issue for it once this one's merged.

I still think we should ensure this implementation is future proof to reduce chances of breaking changes when we implement this.

**Comment by bersalazar** [comment]
> @bersalazar could you add some tests for this feature?

@yxxhero added it

File: pkg/providers/k8s/k8s.go (hunk 2 above)
Comment (bersalazar) [reply pkg/providers/k8s/k8s.go]: Thanks. Separated this.

File: pkg/providers/k8s/k8s.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+func New(l *log.Logger, cfg api.StaticConfig) (*provider, error) {
+	p := &provider{
+		log: l,
+	}
+
+	kubeConfig, err := getKubeConfig(cfg)
+	if err != nil {
+		fmt.Printf("Unable to get a valid Kubeconfig path: %s\n", err)
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go]: why not use p.log to log?

File: pkg/providers/k8s/k8s.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	p.KubeContext = getKubeContext(cfg)
+
+	return p, nil
+}
+
+func getKubeConfig(cfg api.StaticConfig) (string, error) {
+	// Use kubeConfigPath from URI parameters if specified
+	if cfg.String("kubeConfigPath") != "" {
+		if _, err := os.Stat(cfg.String("kubeConfigPath")); err != nil {
+			return cfg.String("kubeConfigPath"), fmt.Errorf("kubeConfigPath URI parameter is set but path %s does not exist.", cfg.String("kubeConfigPath"))
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go]: return "" is ok.

File: pkg/providers/k8s/k8s.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+		}
+	}
+
+	// Use path in KUBECONFIG environment variable if set
+	if envPath := os.Getenv("KUBECONFIG"); envPath != "" {
+		if _, err := os.Stat(envPath); err != nil {
+			return envPath, fmt.Errorf("KUBECONFIG environment variable is set but path %s does not exist.", envPath)
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go]: return ""

File: pkg/providers/k8s/k8s.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+	defaultPath := homeDir + "/.kube/config"
+	if _, err := os.Stat(defaultPath); err == nil {
+		return defaultPath, nil
+	}
+
+	return "", fmt.Errorf("No path was found in any of the following: kubeContext URI param, KUBECONFIG environment variable, or default path %s does not exist.", defaultPath)
+}
+
+func (p *provider) GetString(path string) (string, error) {
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go:72]: add some unit-tests will be better for this func.

File: pkg/providers/k8s/k8s.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+	return string(secret), nil
+}
+
+func (p *provider) GetStringMap(path string) (map[string]interface{}, error) {
+	return nil, fmt.Errorf("This provider does not support values from URI fragments")
+}
+
+// Return an empty Kube context if none is provided
+func getKubeContext(cfg api.StaticConfig) string {
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go:121]: same here.

**Comment by yxxhero** [comment]
@bersalazar please see my newest comments. it will be perfect after you update.

**Comment by bersalazar** [comment]
> > > What about config maps? Maybe we can prep to support them in the future by leaving room in the api for them?
> > > Maybe /[namespace/][apiVersion/kind/]name[#path.to.key]
> > 
> > 
> > Definitely worth considering. Maybe in a future PR, happy to raise an issue for it once this one's merged.
> 
> I still think we should ensure this implementation is future proof to reduce chances of breaking changes when we implement this.

@aslafy-z sure thing, I have modified the proposed API so its more generic and can cover ConfigMaps as well, eventually. Thanks for suggesting! it makes sense to avoid future breaking changes, as you mentioned.

**Comment by bersalazar** [comment]
> @bersalazar please see my newest comments. it will be perfect after you update.

@yxxhero I've had a look at them and pushed updates, including unit tests.

File: README.md [hunk 10]
```diff
@@ -722,6 +724,29 @@ Examples:
...
+
+Environment variables:
+
+- `KUBECONFIG` contains the path to the Kubeconfig that will be used to fetch the secret.
+
+Examples:
+
+- `ref+k8s://v1/Secret/mynamespace/mysecret/foo`
+- `ref+k8s://v1/Secret/mynamespace/mysecret/bar?kubeConfigPath=/home/user/kubeconfig`
+- `secretref+k8s://v1/Secret/mynamespace/secrets/baz`
```
Comment (aslafy-z) [inline README.md]:
```suggestion
- `secretref+k8s://v1/Secret/mynamespace/mysecret/baz`
```

File: README.md [hunk 11]
```diff
@@ -722,6 +724,29 @@ Examples:
...
+- `secretref+k8s://v1/Secret/mynamespace/secrets/baz?kubeContext=minikube`
```
Comment (aslafy-z) [inline README.md]:
```suggestion
- `secretref+k8s://v1/Secret/mynamespace/mysecret/baz?kubeContext=minikube`
```

**Review by aslafy-z — COMMENTED** [review]
Thank you for taking my point into account! Here's two little nit so we make the doc obvious for the secretref case

**Comment by yxxhero** [comment]
@bersalazar please fix lint error.

File: pkg/providers/k8s/k8s.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+	if cfg.String("kubeContext") != "" {
+		return cfg.String("kubeContext")
+	} else {
+		return ""
+	}
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go:124]: if cfg.String("kubeContext") != "" {
		return cfg.String("kubeContext")
	}
	return ""

Comment (bersalazar) [reply pkg/providers/k8s/k8s.go:124]: Thanks! updated.

File: README.md (hunk 11 above)
Comment (bersalazar) [reply README.md]: Thanks, updated.

**Comment by bersalazar** [comment]
> @bersalazar please fix lint error.

@yxxhero all good now, thanks.

Do you mind having a look at https://github.com/helmfile/helmfile/discussions/1212?
It's not directly related to this PR, but it's one of the first use cases for this new k8s provider.
Would appreciate understanding how to deal with passing the kube context down to other helmfiles, or at least see if we need to raise an issue for it.

File: pkg/providers/k8s/k8s.go [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+	return p, nil
+}
+
+func getKubeConfig(cfg api.StaticConfig) (string, error) {
+	// Use kubeConfigPath from URI parameters if specified
+	if cfg.String("kubeConfigPath") != "" {
+		if _, err := os.Stat(cfg.String("kubeConfigPath")); err != nil {
+			return "", fmt.Errorf("kubeConfigPath URI parameter is set but path %s does not exist.", cfg.String("kubeConfigPath"))
+		}
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s.go:46]: if cfg.String("kubeConfigPath") exists, what will happen?

Comment (yxxhero) [reply pkg/providers/k8s/k8s.go:46]: @bersalazar

Comment (bersalazar) [reply pkg/providers/k8s/k8s.go:46]: Updated :+1:

File: pkg/providers/k8s/k8s_test.go [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+			}
+
+			if diff := cmp.Diff(tc.want, got); diff != "" {
+				t.Errorf("unexpected result: -(want), +(got)\n%s", diff)
+			}
+		})
+	}
+}
+
+func Test_getKubeConfig(t *testing.T) {
```
Comment (yxxhero) [inline pkg/providers/k8s/k8s_test.go:87]: @bersalazar I thinks this can't cover all case. WDYT?

Comment (yxxhero) [reply pkg/providers/k8s/k8s_test.go:87]: @bersalazar you can try to use AI to help you generate unit tests. it's so cool.

Comment (bersalazar) [reply pkg/providers/k8s/k8s_test.go:87]: You are correct, tests didn't cover KUBECONFIG env var. I've added a few more tests. Can you please review?

**Pull request merged**